stl/inc/mdspan

```diff
@@ -24,151 +24,126 @@ _STL_DISABLE_CLANG_WARNINGS
 
-    template <class... _IndexTypes>
-        requires (sizeof...(_IndexTypes) == sizeof...(_Extents))
-    constexpr _Mdspan_extent_type(_IndexTypes... /*_OtherExtents*/) noexcept {}
-
-    template <class _OtherIndexType, size_t _Size, size_t... _Idx>
-        requires (_Size == sizeof...(_Extents)) || (_Size == 0)
-    constexpr _Mdspan_extent_type(span<_OtherIndexType, _Size>, index_sequence<_Idx...>) noexcept {}
-
-    constexpr index_type* _Begin_dynamic_extents() noexcept {
-        return nullptr;
+    _NODISCARD static constexpr rank_type _Dynamic_index_inv(rank_type _Idx) noexcept {
+        return _Dynamic_indices_inv[_Idx];
     }
 
-    constexpr const index_type* _Begin_dynamic_extents() const noexcept {
-        return nullptr;
-    }
-};
+    struct _Static_extents_only {
+        constexpr explicit _Static_extents_only() noexcept = default;
 
-template <class _IndexType>
-struct _Mdspan_extent_type<_IndexType, 0> {
-    using index_type = _IndexType;
+        template <class... _Args>
+        constexpr explicit _Static_extents_only(_Args&&...) noexcept {}
 
-    constexpr index_type* _Begin_dynamic_extents() {
-        return nullptr;
-    }
+        _NODISCARD constexpr index_type* begin() const noexcept {
+            return nullptr;
+        }
+    };
 
-    constexpr const index_type* _Begin_dynamic_extents() const noexcept {
-        return nullptr;
-    }
-};
+    static constexpr rank_type _Rank_dynamic = _Dynamic_index(rank());
+    conditional_t<_Rank_dynamic != 0, array<index_type, _Rank_dynamic>, _Static_extents_only> _Dynamic_extents{};
 
-_EXPORT_STD
-template <class _IndexType, size_t... _Extents>
-class extents : private _Mdspan_extent_type<_IndexType, ((_Extents == dynamic_extent) + ... + 0), _Extents...> {
 public:
-    using _Mybase    = _Mdspan_extent_type<_IndexType, ((_Extents == dynamic_extent) + ... + 0), _Extents...>;
-    using index_type = typename _Mybase::index_type;
-    using size_type  = make_unsigned_t<index_type>;
-    using rank_type  = size_t;
-
-    static_assert(_Is_standard_integer<_IndexType>,
-        "IndexType must be a signed or unsigned integer type (N4928 [mdspan.extents.overview]/1.1).");
-    static_assert(((_Extents == dynamic_extent || _STD in_range<_IndexType>(_Extents)) && ...),
-        "Each element of Extents must be either equal to dynamic_extent, or must be representable as a value of type "
-        "IndexType (N4928 [mdspan.extents.overview]/1.2).");
-
-    _NODISCARD static constexpr rank_type rank() noexcept {
-        return sizeof...(_Extents);
-    }
-
     _NODISCARD static constexpr rank_type rank_dynamic() noexcept {
-        return ((_Extents == dynamic_extent) + ... + 0);
+        return _Rank_dynamic;
     }
 
     _NODISCARD static constexpr size_t static_extent(const rank_type _Idx) noexcept {
-        return _Mybase::_Static_extents[_Idx];
+        _STL_VERIFY(_Idx < rank(), "Index must be less than rank() (N4928 [mdspan.extents.obs]/1)");

```

Comment:
`_STL_VERIFY` performs the check all of the time, and these occurrences aren't guarded by IDL or CDL. (We haven't exactly put a lot of design effort into CDL, but the general idea seems to be that it should perform cheap checks that don't require ABI-changing bookkeeping.) Except in highly unusual cases, we prefer release mode with IDL=0 and CDL=0 to perform no precondition checks.

`_STL_ASSERT` is controlled by `_DEBUG` and we use that occasionally, although it's probably inconsistent that elsewhere we use `_STL_VERIFY` guarded by IDL or CDL.